projects/extension/src/content/ExtensionMessageRouter.ts

```diff
@@ -57,11 +57,15 @@ export class ExtensionMessageRouter {
     // forward any messages: extension -> page
     port.onMessage.addListener((data): void => {
       debug(`RECEIVED MESSAGE FROM ${chainId} PORT`, data)
-      sendMessage({
+      const msg = {
         ...data,
         chainId,
         origin: CONTENT_SCRIPT_ORIGIN,
-      })
+      }
+
+      // Because the construction of the message is hacky, we have no choice but to  cast
+      // to `unknown`. This is expected to be fixed in a later refactor.
+      sendMessage(msg as unknown as ToApplication)
```

Comment:
A more accurate way of doing this could have been:

```ts
    type HeaderlessToApplication<T extends ToApplication> = T extends {
      origin: "substrate-connect-extension"                            
    } & infer V                                                        
      ? Omit<V, "chainId">                                             
      : unknown                                                        
    port.onMessage.addListener(                                        
      (data: HeaderlessToApplication<ToApplication>): void => {        
        debug(`RECEIVED MESSAGE FROM ${chainId} PORT`, data)           
        sendMessage({                                                  
          ...data,                                                     
          chainId,                                                     
          origin: CONTENT_SCRIPT_ORIGIN,                               
        })                                                             
      },                                                               
    )                                                                  
```

Reply:
I'm changing that code soon